File: starters/typescript/workflow-runner/src/application/job-factory.ts
/**
 * Job Factory - Creates Workflow Steps
 *
 * Factory functions for creating standardized jobs (workflow steps).
 * Jobs created by this factory are the building blocks of workflows.
 *
 * ## What is a Job?
 * A job is a **workflow step** - a single unit of work that:
 * - Takes input (via context and dependencies)
 * - Performs processing
 * - Returns a result (success/failure with message)
 *
 * ## Responsibilities
 * - Create job instances from configuration
 * - Wire up dependency injection
 * - Automate timing, logging, and error handling
 *
 * ## NOT Responsible For
 * - Workflow orchestration (that's the runner's job)
 * - Retry/skip decisions
 * - Job scheduling
 *
 * ## Design Philosophy
 * Jobs are stateless workflow steps. They don't know about other jobs
 * or their position in a workflow. This makes them reusable and testable.
 *
 * @module application/job-factory
 */

import type { Logger } from "pino";
import type { AwilixContainer } from "awilix";
import { createLogger } from "../infra/logger/pino.js";
import type { JobResult, JobContext } from "../runner/types.js";
import {
  container as defaultContainer,
  type Dependencies,
} from "../di/index.js";

/**
 * Job executor function type (without DI)
 * Receives job context and returns partial result (without executedAt and duration)
 */
export type JobExecutor = (
  ctx: JobContext,
  logger: Logger
) => Promise<Omit<JobResult, "executedAt" | "duration">>;

/**
 * Job executor function type with DI
 * Receives job context, logger, and dependencies
 */
export type JobExecutorWithDeps = (
  ctx: JobContext,
  logger: Logger,
  deps: Dependencies
) => Promise<Omit<JobResult, "executedAt" | "duration">>;

/**
 * Options for creating a job (workflow step)
 */
export interface CreateJobOptions {
  /** Unique job name - identifies this workflow step */
  name: string;
  /** Human-readable description of what this step does */
  description?: string;
  /** Recommended cron schedule (informational only - not enforced) */
  schedule?: string;
  /** Maximum execution time in milliseconds */
  timeout?: number;
}

/**
 * Job definition - a configured workflow step ready for execution
 *
 * This is what the factory returns. The handler can be invoked by the
 * workflow runner to execute this step.
 */
export interface JobDefinition {
  /** Unique identifier for this workflow step */
  name: string;
  /** Human-readable description */
  description: string;
  /** Recommended cron schedule (informational) */
  schedule?: string;
  /** Maximum execution time in milliseconds */
  timeout?: number;
  /** Execute this workflow step (optionally with custom DI container) */
  handler: (container?: AwilixContainer<Dependencies>) => Promise<JobResult>;
}

/**
 * Create a standardized job with automatic timing, logging, and error handling
 *
 * This factory function reduces boilerplate by automatically handling:
 * - Execution timing and duration calculation
 * - Logger creation with job context
 * - Error catching and standardized error responses
 * - Execution ID generation for tracing
 *
 * @param options - Job configuration options
 * @param executor - The actual job logic to execute
 * @returns A job function that can be called or registered
 *
 * @example
 * ```typescript
 * const myJob = createJob(
 *   { name: "my-job", description: "Does something useful" },
 *   async (ctx, logger) => {
 *     logger.info("Starting work...");
 *     // ... do work ...
 *     return {
 *       success: true,
 *       message: "Work completed",
 *     };
 *   }
 * );
 *
 * // Execute the job
 * const result = await myJob.handler();
 * ```
 */
export function createJob(
  options: CreateJobOptions,
  executor: JobExecutor
): JobDefinition {
  const { name, description = "", schedule, timeout } = options;

  const handler = async (
    containerOverride?: AwilixContainer<Dependencies>
  ): Promise<JobResult> => {
    const c = containerOverride ?? defaultContainer;
    const clock = c.resolve("clock");
    const executionIdGen = c.resolve("executionId");

    const startTime = clock.now();
    const executionId = executionIdGen.generate();
    const logger = createLogger({ job: name, executionId });

    const ctx: JobContext = {
      jobName: name,
      executionId,
      startTime,
    };

    logger.info({ executionId }, "Job started");

    try {
      const result = await executor(ctx, logger);

      const duration = clock.now() - startTime;
      logger.info({ duration: `${duration}ms` }, "Job completed");

      return {
        ...result,
        executedAt: new Date(),
        duration,
      };
    } catch (error) {
      const duration = clock.now() - startTime;
      const errorMessage =
        error instanceof Error ? error.message : "Unknown error";

      logger.error(
        {
          error: errorMessage,
          duration: `${duration}ms`,
        },
        "Job failed"
      );

      return {
        success: false,
        message: errorMessage,
        executedAt: new Date(),
        duration,
      };
    }
  };

  return {
    name,
    description,
    schedule,
    timeout,
    handler,
  };
}

/**
 * Create a job with dependency injection support
 *
 * Use this when your job needs access to injectable dependencies
 * like databases, external services, or time functions.
 *
 * @param options - Job configuration options
 * @param executor - The job logic with access to dependencies
 * @returns A job function that can be called with an optional container override
 *
 * @example
 * ```typescript
 * const myJob = createJobWithDeps(
 *   { name: "my-job", description: "Uses DI" },
 *   async (ctx, logger, deps) => {
 *     const { itemRepo, clock, config } = deps;
 *     const items = itemRepo.list();
 *     logger.info({ time: clock.toISOString(), count: items.length }, "Items loaded");
 *     return { success: true, message: "Done" };
 *   }
 * );
 *
 * // Production: uses default container
 * const result = await myJob.handler();
 *
 * // Test: inject mock container
 * const testResult = await myJob.handler(testContainer);
 * ```
 */
export function createJobWithDeps(
  options: CreateJobOptions,
  executor: JobExecutorWithDeps
): JobDefinition {
  const { name, description = "", schedule, timeout } = options;

  const handler = async (
    containerOverride?: AwilixContainer<Dependencies>
  ): Promise<JobResult> => {
    const c = containerOverride ?? defaultContainer;

    // Resolve all dependencies
    const deps: Dependencies = {
      clock: c.resolve("clock"),
      executionId: c.resolve("executionId"),
      config: c.resolve("config"),
      db: c.resolve("db"),
      itemRepo: c.resolve("itemRepo"),
    };

    const startTime = deps.clock.now();
    const executionId = deps.executionId.generate();
    const logger = createLogger({ job: name, executionId });

    const ctx: JobContext = {
      jobName: name,
      executionId,
      startTime,
    };

    logger.info({ executionId }, "Job started");

    try {
      const result = await executor(ctx, logger, deps);

      const duration = deps.clock.now() - startTime;
      logger.info({ duration: `${duration}ms` }, "Job completed");

      return {
        ...result,
        executedAt: new Date(),
        duration,
      };
    } catch (error) {
      const duration = deps.clock.now() - startTime;
      const errorMessage =
        error instanceof Error ? error.message : "Unknown error";

      logger.error(
        {
          error: errorMessage,
          duration: `${duration}ms`,
        },
        "Job failed"
      );

      return {
        success: false,
        message: errorMessage,
        executedAt: new Date(),
        duration,
      };
    }
  };

  return {
    name,
    description,
    schedule,
    timeout,
    handler,
  };
}

File: starters/rust-batch/src/application/job_factory.rs
//! Job Factory - Creates Workflow Steps
//!
//! Factory functions for creating standardized jobs (workflow steps).
//! Jobs created by this factory are the building blocks of workflows.
//!
//! ## What is a Job?
//! A job is a **workflow step** - a single unit of work that:
//! - Takes input (via context and dependencies)
//! - Performs processing
//! - Returns a result (success/failure with message)
//!
//! ## Responsibilities
//! - Create job instances from configuration
//! - Wire up dependency injection
//! - Automate timing, logging, and error handling

use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use tracing::{error, info, Span};

use crate::application::job_types::{JobContext, JobResult, PartialJobResult};
use crate::di::AppContext;
use crate::shared::Timer;

/// Job executor function type
///
/// Receives job context and app context, returns partial result
pub type JobExecutor = Arc<
    dyn Fn(
            JobContext,
            Arc<AppContext>,
        ) -> Pin<Box<dyn Future<Output = Result<PartialJobResult, anyhow::Error>> + Send>>
        + Send
        + Sync,
>;

/// Options for creating a job (workflow step)
#[derive(Debug, Clone)]
pub struct JobOptions {
    /// Unique job name - identifies this workflow step
    pub name: String,
    /// Human-readable description of what this step does
    pub description: String,
    /// Recommended cron schedule (informational only - not enforced)
    pub schedule: Option<String>,
    /// Maximum execution time in milliseconds
    pub timeout_ms: Option<u64>,
}

impl JobOptions {
    pub fn new(name: impl Into<String>, description: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            description: description.into(),
            schedule: None,
            timeout_ms: None,
        }
    }

    pub fn with_schedule(mut self, schedule: impl Into<String>) -> Self {
        self.schedule = Some(schedule.into());
        self
    }

    pub fn with_timeout(mut self, timeout_ms: u64) -> Self {
        self.timeout_ms = Some(timeout_ms);
        self
    }
}

/// Job definition - a configured workflow step ready for execution
///
/// This is what the factory returns. The handler can be invoked by the
/// workflow runner to execute this step.
#[derive(Clone)]
pub struct JobDefinition {
    /// Unique identifier for this workflow step
    pub name: String,
    /// Human-readable description
    pub description: String,
    /// Recommended cron schedule (informational)
    pub schedule: Option<String>,
    /// Maximum execution time in milliseconds
    pub timeout_ms: Option<u64>,
    /// The executor function
    executor: JobExecutor,
}

impl JobDefinition {
    /// Execute this workflow step
    pub async fn execute(&self, ctx: Arc<AppContext>) -> JobResult {
        let timer = Timer::new();
        let execution_id = ctx.execution_id.generate();

        let job_ctx = JobContext {
            job_name: self.name.clone(),
            execution_id: execution_id.clone(),
            start_time: ctx.clock.now(),
        };

        let span = Span::current();
        span.record("job", &self.name);
        span.record("execution_id", &execution_id);

        info!(
            job = %self.name,
            execution_id = %execution_id,
            "Job started"
        );

        let result = (self.executor)(job_ctx, ctx.clone()).await;

        let duration_ms = timer.elapsed_ms();
        let executed_at = ctx.clock.now_utc();

        match result {
            Ok(partial) => {
                info!(
                    job = %self.name,
                    duration_ms = %duration_ms,
                    success = %partial.success,
                    "Job completed"
                );
                partial.finalize(executed_at, duration_ms)
            }
            Err(e) => {
                error!(
                    job = %self.name,
                    duration_ms = %duration_ms,
                    error = %e,
                    "Job failed"
                );
                JobResult {
                    success: false,
                    message: e.to_string(),
                    executed_at,
                    duration_ms,
                    metrics: None,
                    details: None,
                }
            }
        }
    }
}

/// Create a standardized job with automatic timing, logging, and error handling
///
/// # Example
///
/// ```ignore
/// let job = create_job(
///     JobOptions::new("hello-world", "A simple hello world job"),
///     |ctx, app_ctx| Box::pin(async move {
///         Ok(JobResult::success("Hello, World!"))
///     })
/// );
/// ```
pub fn create_job<F, Fut>(options: JobOptions, executor: F) -> JobDefinition
where
    F: Fn(JobContext, Arc<AppContext>) -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<PartialJobResult, anyhow::Error>> + Send + 'static,
{
    let executor = Arc::new(move |ctx: JobContext, app_ctx: Arc<AppContext>| {
        let fut = executor(ctx, app_ctx);
        Box::pin(fut)
            as Pin<Box<dyn Future<Output = Result<PartialJobResult, anyhow::Error>> + Send>>
    });

    JobDefinition {
        name: options.name,
        description: options.description,
        schedule: options.schedule,
        timeout_ms: options.timeout_ms,
        executor,
    }
}
